src/analyzer/protocol/postgresql/postgresql.spicy

```diff
@@ -0,0 +1,335 @@
+        } else {
+            # print "frontend ssl_state backend set!", self.context();
+            if (!self.s1_connected) {
+                if (self.context().ssl_backend_state == SSLBackendState::S) {
+                    if (!self.context().ssl_sink_connected) {
+                        self.context().ssl_sink.connect(new SSLSink());
+                        self.context().ssl_sink_connected = True;
+                    }
+                } else {
+                    # print "connecting plain frontend messages";
+                    self.s1.connect(new PlainFrontendMessages());
+                }
+
+                self.s1_connected = True;
+
+                if (|self.buffered| > 0) {
+                    for (b in self.buffered)
+                        self.s1.write(b);
+                }
+
+                self.buffered.resize(0);
+            }
+
+            self.s1.write($$);
+        }
+    }
+};
+
+type PlainFrontendMessages = unit {
+    startup_message: StartupMessage;
+    : FrontendMessage[];
+};
+
+type FrontendMessage = unit {
+    typ: uint8;
+    length: uint32 &requires=(self.length >= 4);
+
+    switch (self.typ) {
+        'p' -> : AuthenticationResponse(self.length - 4) &size=self.length - 4;
+        'X' -> : Terminate(self.length - 4) &size=self.length - 4;
+        'Q' -> simple_query: SimpleQuery(self.length - 4) &size=self.length - 4;
+        * -> not_implemented: NotImplemented(self.typ, self.length - 4) &size=self.length - 4;
+    };
+};
+
+type AuthenticationResponse = unit(length: uint32) {
+    # This is PasswordMessage, SASLInitialMessage, etc. based on context.
+    # For now, just thread it through.
+    data: bytes &eod;
+};
+
+type Terminate = unit(length: uint32) {};
+
+type SimpleQuery = unit(length: uint32) {
+    query: bytes &until=b"\x00" &size=length;
+};
+
+# The client has requested SSL, the server either confirms (S) or
+# stays in plaintext (N) mode. Depending on the result, we connect
+# our sink to the SSL sink, or to a PlainBackendMessages unit.
+#
+type MaybeBackendSSL = unit(ctx: Context&) {
+
+    # Connected SSL, or plaintext.
+    sink s1;
+
+    ssl_byte: uint8 &requires=($$ == 'S' || $$ == 'N') {
+        # print "backend ssl_byte", $$;
+        if ($$ == 'S') {
+            ctx.ssl_backend_state = SSLBackendState::S;
+            if (!ctx.ssl_sink_connected) {
+                ctx.ssl_sink.connect(new SSLSink());
+                ctx.ssl_sink_connected = True;
+            }
+
+            # Share the SSL sink with the frontend.
+            self.s1 = ctx.ssl_sink;
+        } else {
+            ctx.ssl_backend_state = SSLBackendState::N;
+            self.s1.connect(new PlainBackendMessages());
+        }
+    }
+
+    # Now that s1 is connected, forward the rest of the connection to it.
+    : bytes &chunked &eod -> self.s1;
+};
+
+# Top-level entry for the server.
+public type BackendMessages = unit {
+    %context = Context;
+
+    var buffered: vector<bytes>;
```

Comment:
Great approach to avoid the naive solution of appending to one big `bytes` (which could require expensive reallocs at runtime).